smb_streams.c

```diff
@@ -63,6 +64,112 @@ typedef struct _php_smb_stream_data {
 	smbc_lseek_fn           smbc_lseek;
 } php_smb_stream_data;
 
+struct _php_smb_pool {
+	unsigned char          hash[20];
+	php_smbclient_state    *state;
+	struct _php_smb_pool   *next;
+	int                    nb;
+};
+
+struct _php_smb_pool *first = NULL;
+
+
+static php_smbclient_state *php_smb_pool_get(php_stream_context *context, const char *url TSRMLS_DC)
+{
+	PHP_SHA1_CTX          sha1;
+	unsigned char         hash[20];
+	struct _php_smb_pool *pool;
+
+	/* Create a hash for connection parameter */
+	PHP_SHA1Init(&sha1);
+	if (!memcmp(url, "smb://", 6)) {
+		char *p;
+		p = strchr(url+6, '/'); // we only want smb://workgroup;user:pass@server/
+		PHP_SHA1Update(&sha1, (const unsigned char *)url+6, p ? p - url - 6 : strlen(url+6));
+	}
+	if (context) {
+#if PHP_MAJOR_VERSION >= 7
+		zval *tmpzval;
+
+		if (NULL != (tmpzval = php_stream_context_get_option(context, "smb", "workgroup"))) {
+			if (Z_TYPE_P(tmpzval) == IS_STRING) {
+				PHP_SHA1Update(&sha1, (const unsigned char *)Z_STRVAL_P(tmpzval), Z_STRLEN_P(tmpzval)+1);
+			}
+		}
+		if (NULL != (tmpzval = php_stream_context_get_option(context, "smb", "username"))) {
+			if (Z_TYPE_P(tmpzval) == IS_STRING) {
+				PHP_SHA1Update(&sha1, (const unsigned char *)Z_STRVAL_P(tmpzval), Z_STRLEN_P(tmpzval)+1);
+			}
+		}
+		if (NULL != (tmpzval = php_stream_context_get_option(context, "smb", "password"))) {
+			if (Z_TYPE_P(tmpzval) == IS_STRING) {
+				PHP_SHA1Update(&sha1, (const unsigned char *)Z_STRVAL_P(tmpzval), Z_STRLEN_P(tmpzval)+1);
+			}
+		}
+#else
+		zval **tmpzval;
+
+		if (php_stream_context_get_option(context, "smb", "workgroup", &tmpzval) == SUCCESS) {
+			if (Z_TYPE_PP(tmpzval) == IS_STRING) {
+				PHP_SHA1Update(&sha1, (const unsigned char *)Z_STRVAL_PP(tmpzval), Z_STRLEN_PP(tmpzval)+1);
+			}
+		}
+		if (php_stream_context_get_option(context, "smb", "username", &tmpzval) == SUCCESS) {
+			if (Z_TYPE_PP(tmpzval) == IS_STRING) {
+				PHP_SHA1Update(&sha1, (const unsigned char *)Z_STRVAL_PP(tmpzval), Z_STRLEN_PP(tmpzval)+1);
+			}
+		}
+		if (php_stream_context_get_option(context, "smb", "password", &tmpzval) == SUCCESS) {
+			if (Z_TYPE_PP(tmpzval) == IS_STRING) {
+				PHP_SHA1Update(&sha1, (const unsigned char *)Z_STRVAL_PP(tmpzval), Z_STRLEN_PP(tmpzval)+1);
+			}
+		}
+#endif
+	}
+	PHP_SHA1Final(hash, &sha1);
+
+	/* Reuse state from pool if exists */
+	for (pool=first; pool ; pool=pool->next) {
+		if (!memcmp(hash, pool->hash, 20)) {
+			pool->nb++;
+			return pool->state;
+		}
+	}
+
+	/* Crate a new state and save it in the pool */
+	pool = emalloc(sizeof(*pool));
+	memcpy(pool->hash, hash, 20);
+	pool->nb    = 1;
+	pool->next  = first;
```

Comment:
Probably shouldn't be aligning the equals (no extra spaces).
Sorry, I've worked with some picky people, but again, no big deal :)